### PR TITLE
#728 - Add support for Spring WebFlux.

### PR DESCRIPTION
**Pull request by gregturn**
* Render hypermedia from WebFlux controllers.
* Support consuming hypermedia through WebClient.
* Support link creation reactively with a custom filter and explicitly with a user-provided ServerWebExchange.
* Add a ReactiveResourceAssembler and SimpleReactiveResourceAssembler.
* Upgrade to Spring Framework 5.2 to align with Spring Data Moore, removing specialized Forwarded header handling.

### CODE REVIEW

**Comment by gregturn** [comment]
@schauder @odrotbohm This PR contains a lot. I did move some things around so they could be reused by both Spring MVC and by WebFlux. I tried to contain reactor-specific in subpackages as much as possible.

This PR definitely needs to be followed by some documentation upgrades.

File: src/main/java/org/springframework/hateoas/config/reactive/WebClientConfigurer.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+		if (this.hypermediaTypes.contains(HypermediaType.UBER)) {
+
+			ObjectMapper uberObjectMapper = createUberObjectMapper(this.mapper);
+
+			encoders.add(new Jackson2JsonEncoder(uberObjectMapper, MediaTypes.UBER_JSON));
+			decoders.add(new Jackson2JsonDecoder(uberObjectMapper, MediaTypes.UBER_JSON));
+		}
+
+		// Include ability to decode into a plain string
+		decoders.add(StringDecoder.allMimeTypes());
```
[the remark this answers is not in the document]
Comment (gregturn) [reply src/main/java/org/springframework/hateoas/config/reactive/WebClientConfigurer.java]: For that to occur someone must provide an accept header for JPG AND request it be converted to a string. 

I frankly think the proper solution is a revision to Spring Framework’s default encoders such that I don’t have to disable them.

File: src/main/java/org/springframework/hateoas/config/mvc/HypermediaWebMvcConfigurer.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+			.map(AbstractJackson2HttpMessageConverter::getObjectMapper)
+			.anyMatch(Jackson2HalModule::isAlreadyRegisteredIn)) {
+
+			return;
+		}
+
+		MessageSourceAccessor linkRelationMessageSource = this.beanFactory.getBean(MESSAGE_SOURCE_BEAN_NAME,
+				MessageSourceAccessor.class);
+
+		if (this.hypermediaTypes.stream().anyMatch(HypermediaType::isHalBasedMediaType)) {
```
[the remark this answers is not in the document]
Comment (gregturn) [reply src/main/java/org/springframework/hateoas/config/mvc/HypermediaWebMvcConfigurer.java]: Not sure performance is a big issue here. But I wish we had a more elegant check. 

Actually I’m not 100% certain we need this check anymore given how I’ve reareanged the registrar.

[the remark this answers is not in the document]
Comment (gregturn) [reply src/main/java/org/springframework/hateoas/config/mvc/HypermediaWebMvcConfigurer.java]: Actually I don’t think the HAL check is needed anymore. I was removing it in my new feature to register new media types.

File: src/main/java/org/springframework/hateoas/config/reactive/HypermediaWebFluxConfigurer.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+			// Already configured for hypermedia!
+			return;
+		}
+
+		MessageSourceAccessor linkRelationMessageSource = this.beanFactory.getBean(MESSAGE_SOURCE_BEAN_NAME,
+			MessageSourceAccessor.class);
+
+		CodecConfigurer.CustomCodecs customCodecs = configurer.customCodecs();
+
+		if (this.hypermediaTypes.stream().anyMatch(HypermediaType::isHalBasedMediaType)) {
```
[the remark this answers is not in the document]
Comment (gregturn) [reply src/main/java/org/springframework/hateoas/config/reactive/HypermediaWebFluxConfigurer.java]: Again not sure that check gives us anything anymore.

File: src/main/java/org/springframework/hateoas/config/reactive/WebClientConfigurer.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	 */
+	public ExchangeStrategies hypermediaExchangeStrategies() {
+
+		MessageSourceAccessor linkRelationMessageSource = this.beanFactory.getBean(MESSAGE_SOURCE_BEAN_NAME,
+			MessageSourceAccessor.class);
+
+		List<Encoder<?>> encoders = new ArrayList<>();
+		List<Decoder<?>> decoders = new ArrayList<>();
+
+		if (this.hypermediaTypes.stream().anyMatch(HypermediaType::isHalBasedMediaType)) {
```
[the remark this answers is not in the document]
Comment (gregturn) [reply src/main/java/org/springframework/hateoas/config/reactive/WebClientConfigurer.java]: Same.

File: src/test/resources/logback.xml [hunk 5]
```diff
@@ -7,7 +7,13 @@
 		</encoder>
 	</appender>
 
-	<logger name="org.springframework.hateoas" level="error" />
+	<!--<logger name="org.springframework.hateoas" level="error" />-->
+	<logger name="org.springframework.hateoas" level="trace" />
```
[the remark this answers is not in the document]
Comment (gregturn) [reply src/test/resources/logback.xml]: Left over.

File: src/main/java/org/springframework/hateoas/config/reactive/HypermediaWebFluxConfigurer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (schauder) [reply src/main/java/org/springframework/hateoas/config/reactive/HypermediaWebFluxConfigurer.java]: I first thought it is completely superfluous, so if you are not concerned about performance, we should remove it.

[the remark this answers is not in the document]
Comment (gregturn) [reply src/main/java/org/springframework/hateoas/config/reactive/HypermediaWebFluxConfigurer.java]: Done.

**Comment by gregturn** [comment]
Resolved via https://github.com/spring-projects/spring-hateoas/commit/de0f2fc09a127ea315a0a5e12f0bd0f9130b8550, https://github.com/spring-projects/spring-hateoas/commit/e887f06f466e72aad2fcfd6f0046983f9216dc10, and https://github.com/spring-projects/spring-hateoas/commit/0d55afa202cc03af4a1875551526223a79825263.

**Pull request closed without merging**